src/include/duckdb/storage/compression/alp/algorithm/alp.hpp

```diff
@@ -0,0 +1,413 @@
+
+		// Save k' best combinations
+		for (idx_t i = 0; i < MinValue(AlpConstants::MAX_COMBINATIONS, (uint8_t)best_k_combinations.size()); i++) {
+			state.best_k_combinations.push_back(best_k_combinations[i]);
+		}
+	}
+
+	/*
+	 * Find the best combination of factor-exponent for a vector from within the best k combinations
+	 * This is ALP second level sampling
+	 */
+	static void FindBestFactorAndExponent(const vector<T> &input_vector, idx_t n_values, State &state) {
+		//! We sample equidistant values within a vector; to do this we skip a fixed number of values
+		vector<T> vector_sample;
+		uint32_t idx_increments = MaxValue(1, (int32_t)std::ceil((double)n_values / AlpConstants::SAMPLES_PER_VECTOR));
+		for (idx_t i = 0; i < n_values; i += idx_increments) {
+			vector_sample.push_back(input_vector[i]);
+		}
+
+		AlpEncodingIndices best_encoding_indices = {0, 0};
+		uint64_t best_total_bits = NumericLimits<uint64_t>::Maximum();
+		idx_t worse_total_bits_counter = 0;
+
+		//! We try each K combination in search for the one which minimize the compression size in the vector
+		for (auto &combination : state.best_k_combinations) {
+			uint64_t estimated_compression_size =
+			    DryCompressToEstimateSize<false>(vector_sample, combination.encoding_indices);
+
+			// If current compression size is worse (higher) or equal than the current best combination
+			if (estimated_compression_size >= best_total_bits) {
+				worse_total_bits_counter += 1;
+				// Early exit strategy
+				if (worse_total_bits_counter == AlpConstants::SAMPLING_EARLY_EXIT_THRESHOLD) {
+					break;
+				}
+				continue;
+			}
+			// Otherwise we replace the best and continue trying with the next combination
+			best_total_bits = estimated_compression_size;
+			best_encoding_indices = combination.encoding_indices;
+			worse_total_bits_counter = 0;
+		}
+		state.vector_encoding_indices = best_encoding_indices;
+	}
+
+	/*
+	 * ALP Compress
+	 */
+	static void Compress(const vector<T> &input_vector, idx_t n_values, const vector<uint16_t> &vector_null_positions,
+	                     idx_t nulls_count, State &state) {
+		if (state.best_k_combinations.size() > 1) {
+			FindBestFactorAndExponent(input_vector, n_values, state);
+		} else {
+			state.vector_encoding_indices = state.best_k_combinations[0].encoding_indices;
+		}
+
+		// Encoding Floating-Point to Int64
+		//! We encode all the values regardless of their correctness to recover the original floating-point
+		//! We detect exceptions later using a predicated comparison
+		vector<T> tmp_decoded_values(n_values, 0); // Tmp array to check wether the encoded values are exceptions
+		for (idx_t i = 0; i < n_values; i++) {
+			T value = input_vector[i];
+			int64_t encoded_value = EncodeValue(value, state.vector_encoding_indices);
+			T decoded_value = DecodeValue(encoded_value, state.vector_encoding_indices);
+			state.encoded_integers[i] = encoded_value;
+			tmp_decoded_values[i] = decoded_value;
```

Comment:
Yes. In this context it makes much more sense to do it this way.
Originally, we did it like this because in the second loop with all the array loaded we can use SIMD to do the comparisons faster. But since we are not using SIMD in this implementation, we have implemented your suggestion.
With this change, the `tmp_decoded_values` array is not needed anymore (solving also the previous comment)